src/Passes/Source/Rules/Factory.cpp

```diff
@@ -0,0 +1,327 @@
+                replacements.push_back({llvm::dyn_cast<Instruction>(val), instr});
+
+                // Deleting the getelement and cast operations
+                replacements.push_back({llvm::dyn_cast<Instruction>(cap["getElement"]), nullptr});
+                replacements.push_back({llvm::dyn_cast<Instruction>(cap["cast"]), nullptr});
+
+                return true;
+            };
+
+        auto get_element  = call("__quantum__rt__array_get_element_ptr_1d", "arrayName"_cap = _, "index"_cap = _);
+        auto cast_pattern = bitCast("getElement"_cap = get_element);
+        auto load_pattern = load("cast"_cap = cast_pattern);
+
+        addRule({std::move(load_pattern), access_replacer});
+
+        /// Release replacement
+        auto deleter = deleteInstruction();
+        addRule(
+            {call("__quantum__rt__qubit_release_array", "name"_cap = _),
+             [qubit_alloc_manager, deleter](Builder& builder, Value* val, Captures& cap, Replacements& rep) {
+                 qubit_alloc_manager->release(cap["name"]->getName().str());
+                 return deleter(builder, val, cap, rep);
+             }
+
+            });
+    }
+
+    void RuleFactory::useStaticQubitAllocation()
+    {
+        auto qubit_alloc_manager = qubit_alloc_manager_;
+        auto allocation_replacer =
+            [qubit_alloc_manager](Builder& builder, Value* val, Captures&, Replacements& replacements) {
+                // Getting the type pointer
+                auto ptr_type = llvm::dyn_cast<llvm::PointerType>(val->getType());
+                if (ptr_type == nullptr)
+                {
+                    return false;
+                }
+
+                // Computing the index by getting the current index value and offseting by
+                // the offset at which the qubit array is allocated.
+                auto offset = qubit_alloc_manager->allocate();
+
+                // Creating a new index APInt that is shifted by the offset of the allocation
+                // TODO(tfr): Get the bitwidth size from somewhere
+                auto idx = llvm::APInt(64, offset);
+
+                // Computing offset
+                auto new_index = llvm::ConstantInt::get(builder.getContext(), idx);
+
+                // TODO(tfr): Understand what the significance of the addressspace is in relation to the
+                // QIR. Activate by uncommenting:
+                // ptr_type = llvm::PointerType::get(ptr_type->getElementType(), 2);
+                auto instr = new llvm::IntToPtrInst(new_index, ptr_type);
+                instr->takeName(val);
+
+                // Replacing the instruction with new instruction
+                replacements.push_back({llvm::dyn_cast<Instruction>(val), instr});
+
+                return true;
+            };
+        addRule({call("__quantum__rt__qubit_allocate"), allocation_replacer});
+
+        // Removing release calls
+        removeFunctionCall("__quantum__rt__qubit_release");
```

Comment:
In the QIR qubit array allocation replacement, the `allocation_replacer` includes a call to the `qubit_alloc_manager->allocate();` method and a corresponding call to `qubit_alloc_manager->release()` in the lambda for removing the QIR qubit array release instructions. Here the `allocation_replacer` still calls allocate on the qubit manager but the removal here of the release instruction doesn't call `qubit_alloc_manager->release()`. Is this because the array allocation is done with a name? The asymmetry seems odd, so I want to make sure it's intentional and not a subtle bug.

Reply:
Yes, this is because of the name. The allocation manager will throw if the same name is allocated twice within the same scope and hence you need to "release" it. I agree that the asymmetry is odd and, in fact, the cleaner thing to do would be to also call release (which would not do anything at this point).